source/celview.cpp

```diff
@@ -136,73 +138,149 @@ void CelView::framePixelClicked(unsigned x, unsigned y)
+void CelView::removeFrames(int startingIndex, int endingIndex)
 {
-    QImage image = QImage(imagefilePath);
+    int idx = startingIndex;
+    while (startingIndex != endingIndex) {
+        this->removeCurrentFrame(idx);
+        startingIndex++;
+    }
+}
 
-    if (image.isNull()) {
+void CelView::sendAddFrameCmd(IMAGE_FILE_MODE mode, int index, const QString &imagefilePath)
+{
+    AddFrameCommand *command;
+    try {
+        command = new AddFrameCommand(mode, index, imagefilePath);
+    } catch (...) {
+        QMessageBox::critical(this, "Error", "Failed to read image file: " + imagefilePath);
+        delete command;
         return;
     }
 
-    this->gfx->replaceFrame(this->currentFrameIndex, image);
+    // send a command to undostack, making adding frame undo/redoable
+    QObject::connect(command, &AddFrameCommand::added, this, &CelView::insertFrames);
+    QObject::connect(command, &AddFrameCommand::undoAdded, this, &CelView::removeFrames);
+
+    undoStack->push(command);
+}
+
+void CelView::replaceCurrentFrame(int frameIdx, const QImage &image)
+{
+    this->gfx->replaceFrame(frameIdx, image);
 
     // update the view
     this->initialize(this->gfx);
     this->displayFrame();
 }
 
-void CelView::removeCurrentFrame()
+void CelView::sendReplaceCurrentFrameCmd(const QString &imagefilePath)
+{
+    QImage image = QImage(imagefilePath);
+
+    if (image.isNull()) {
+        return;
+    }
+
+    // send a command to undostack, making replacing frame undo/redoable
+    ReplaceFrameCommand *command = new ReplaceFrameCommand(this->currentFrameIndex, image, this->gfx->getFrameImage(this->currentFrameIndex));
+    QObject::connect(command, &ReplaceFrameCommand::replaced, this, &CelView::replaceCurrentFrame);
+    QObject::connect(command, &ReplaceFrameCommand::undoReplaced, this, &CelView::replaceCurrentFrame);
+
+    undoStack->push(command);
+}
+
+void CelView::sendRemoveFrameCmd()
+{
+    // send a command to undostack, making deleting frame undo/redoable
+    RemoveFrameCommand *command = new RemoveFrameCommand(this->currentFrameIndex, this->gfx->getFrameImage(this->currentFrameIndex));
+    QObject::connect(command, &RemoveFrameCommand::removed, this, &CelView::removeCurrentFrame);
+    QObject::connect(command, &RemoveFrameCommand::inserted, this, &CelView::insertImageFile);
+
+    this->undoStack->push(command);
+}
+
+void CelView::removeCurrentFrame(int frameIdx)
 {
     // remove the frame
-    this->gfx->removeFrame(this->currentFrameIndex);
+    int removedGroupIdx = this->gfx->removeFrame(frameIdx);
+
+    // if any group index has been removed, we have to store it so we can restore the group if we start inserting back frames
+    // otherwise we need to store current group index, so if user does undo operation on remove frame operation
+    // we can assign the frame to the correct group, where it was before
+    removedFrameGroupIdxs.push(this->currentGroupIndex);
```

Comment:
You told me about this one, but I'll add it to my review anyway so we don't forget. We're pushing to the stack twice here. Need to remove this line.

```suggestion
```

Reply:
Done.